### PR TITLE
[sgd] Update tf logging

### PR DESCRIPTION
**Pull request by maximsmol**
## Why are these changes needed?
Previously the worker nodes would dump 3 separate progress bars to stdout, which looks awful compared to the default tensorflow Progbar.

The cifar example had some issues like being flaky with `--use-gpu`. It also did not provide any comparison between training locally and on distributed workers.

The new implementation will send all progress updates to the head node every `n` steps and display one progress bar instead.

The updates to cifar fix the `--use-gpu` flag and provide a simple benchmark with `--time-only`.

## Related issue number
N/A

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

**Comment by maximsmol** [comment]
Seems like `on_batch_end` is not returning logs appropriately when training with tune, will investigate

**Comment by maximsmol** [comment]
The reason is that `tf_trainer` relies on there being an accuracy and a loss metric, revising

File: python/ray/experimental/sgd/tf/examples/cifar_tf_example.py [hunk 1]
```diff
@@ -182,46 +185,117 @@ def data_augmentation_creator(config):
...
+    def init_hook():
+        os.environ["TF_CPP_MIN_LOG_LEVEL"] = "3"
+        os.environ["TF_CPP_MIN_VLOG_LEVEL"] = "3"
+
+        import tensorflow as tf
+
+        gpus = tf.config.experimental.list_physical_devices("GPU")
+        if len(gpus) == 0:
+            if args.use_gpu:
+                raise "No GPUs found!"
```
Comment (richardliaw) [inline python/ray/experimental/sgd/tf/examples/cifar_tf_example.py]: ValueError?

File: python/ray/experimental/sgd/tf/examples/cifar_tf_example.py [hunk 2]
```diff
@@ -182,46 +185,117 @@ def data_augmentation_creator(config):
...
+
+        import logging
+        tf.get_logger().setLevel(logging.ERROR)
```
Comment (richardliaw) [inline python/ray/experimental/sgd/tf/examples/cifar_tf_example.py]: Can you add comments here about what you're doing?

File: python/ray/experimental/sgd/tf/examples/cifar_tf_example.py [hunk 3]
```diff
@@ -182,46 +185,117 @@ def data_augmentation_creator(config):
...
-    model.fit(dataset, steps_per_epoch=num_train_steps, epochs=1)
-    dt = (time.time() - training_start)
-    print(f"Training on workers takes: {dt:.3f} seconds/epoch")
-
-    scores = model.evaluate(test_dataset, steps=num_eval_steps)
-    print("Test loss:", scores[0])
-    print("Test accuracy:", scores[1])
+        config=config)
+
+    if args.time_only:
```
Comment (richardliaw) [inline python/ray/experimental/sgd/tf/examples/cifar_tf_example.py]: Should this be a separate script? 

The cifar tf example should be a dead simple example that works out of the box (without bells and whistles).

File: python/ray/experimental/sgd/tf/examples/cifar_tf_example.py [hunk 4]
```diff
@@ -182,46 +185,117 @@ def data_augmentation_creator(config):
...
+            verbose=True)
+
+        local_training_time = time.time() - training_start
+
+        local_eval_stats = model.evaluate(
+            test_data, steps=config["evaluate_config"]["steps"], verbose=True)
+
+        print(f"Remote results: {remote_train_stats}:")
+        print(f"Local results. Train: {local_train_stats.history}" +
+              f" Eval: {local_eval_stats}")
```
Comment (richardliaw) [inline python/ray/experimental/sgd/tf/examples/cifar_tf_example.py]: btw, fstrings I think don't work on py3.5

File: python/ray/experimental/sgd/tf/examples/tf-example-sgd.yaml [hunk 5]
```diff
@@ -51,6 +51,8 @@ setup_commands:
     - ray &> /dev/null || pip install -U https://s3-us-west-2.amazonaws.com/ray-wheels/latest/ray-0.9.0.dev0-cp36-cp36m-manylinux1_x86_64.whl
     - pip install -U ray[tune]
     - pip install tensorflow-gpu==2.1.0
+    - sudo rm /usr/local/cuda
```
Comment (richardliaw) [inline python/ray/experimental/sgd/tf/examples/tf-example-sgd.yaml]: Can you add a comment here about what you're doing?

File: python/ray/experimental/sgd/tf/tf_runner.py [hunk 6]
```diff
@@ -16,10 +17,54 @@ def _try_import_strategy():
...
+def _noop():
+    pass
+
+
+def _load_proper_LambdaCallback():
+    # default TF2 LambdaCallback does not support the new callback naming
+
+    import tensorflow as tf
+
+    class LambdaCallback(tf.keras.callbacks.LambdaCallback):
```
Comment (richardliaw) [inline python/ray/experimental/sgd/tf/tf_runner.py]: Is it possible to just use the basic Callback function here?

File: python/ray/experimental/sgd/tf/tf_runner.py [hunk 7]
```diff
@@ -35,15 +80,29 @@ def __init__(self, model_creator, data_creator, config=None,
...
+        if self.init_hook is not None:
+            self.init_hook()
+
         logger.debug("Creating dataset")
         self.train_dataset, self.test_dataset = self.data_creator(self.config)
 
         logger.debug("Creating model")
         self.model = self.model_creator(self.config)
 
+        self._LambdaCallback = _load_proper_LambdaCallback()
```
Comment (richardliaw) [inline python/ray/experimental/sgd/tf/tf_runner.py]: self._callback

File: python/ray/experimental/sgd/tf/tf_runner.py [hunk 8]
```diff
@@ -35,15 +80,29 @@ def __init__(self, model_creator, data_creator, config=None,
...
+        self._recorded_all_results = threading.Event()
+        self._ready_to_continue = threading.Barrier(2)  # main + model
+        self._step_counter = 0
+        self._step_limit = 1
+
+        self._LambdaCallback = None
 
     def setup(self):
         """Initializes the model."""
+        if self.init_hook is not None:
```
Comment (richardliaw) [inline python/ray/experimental/sgd/tf/tf_runner.py]: Can this be called at the very beginning of `__init__`?

File: python/ray/experimental/sgd/tf/tf_trainer.py [hunk 9]
```diff
@@ -92,24 +95,65 @@ def __init__(self,
                 for i, worker in enumerate(self.workers)
             ])
 
-    def train(self):
+        import tensorflow as tf
+        self._Progbar = tf.keras.utils.Progbar
```
Comment (richardliaw) [inline python/ray/experimental/sgd/tf/tf_trainer.py]: do we need self._Progbar? would it be possible to import in _generic_model_driver

File: python/ray/experimental/sgd/tf/tf_trainer.py [hunk 10]
```diff
@@ -92,24 +95,65 @@ def __init__(self,
...
         # see ./tf_runner.py:setup_distributed
         # for an explanation of only taking the first worker's data
-        worker_stats = ray.get([w.step.remote() for w in self.workers])
         stats = worker_stats[0].copy()
+
         return stats
 
-    def validate(self):
+    def train(self, progress_report_interval=1):
+        """Runs a training epoch."""
```
Comment (richardliaw) [inline python/ray/experimental/sgd/tf/tf_trainer.py]: docstring for progress_report_interval?

File: python/ray/experimental/sgd/tf/tf_trainer.py [hunk 11]
```diff
@@ -92,24 +95,65 @@ def __init__(self,
...
+
+    def _generic_model_driver(self, request_step, steps, metrics_prefix):
         """Runs a training epoch."""
+        progbar = self._Progbar(steps)
+
+        worker_stats = [None] * len(self.workers)
+        done = False
+        while not done:
+            reqs = [request_step(w) for w in self.workers]
+            for workerN, (type, logs) in enumerate(ray.get(reqs)):
```
Comment (richardliaw) [inline python/ray/experimental/sgd/tf/tf_trainer.py]: instead of `workerN`, can you do something like `worker_idx` or `worker_index`? Same goes for batch.

File: python/ray/experimental/sgd/tf/tf_trainer.py [hunk 12]
```diff
@@ -92,24 +95,65 @@ def __init__(self,
...
+                if type == "batch":
```
Comment (richardliaw) [inline python/ray/experimental/sgd/tf/tf_trainer.py]: type is a keyword in python; use something like metric_freq?

File: python/ray/experimental/sgd/tf/tf_trainer.py [hunk 13]
```diff
@@ -92,24 +95,65 @@ def __init__(self,
                 for i, worker in enumerate(self.workers)
             ])
 
-    def train(self):
+        import tensorflow as tf
+        self._Progbar = tf.keras.utils.Progbar
+
+    def _generic_model_driver(self, request_step, steps, metrics_prefix):
         """Runs a training epoch."""
```
Comment (richardliaw) [inline python/ray/experimental/sgd/tf/tf_trainer.py]: Can you also add docstring? https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html

File: python/ray/experimental/sgd/tf/tf_runner.py [hunk 14]
```diff
@@ -83,48 +146,119 @@ def setup_distributed(self, urls, world_rank, world_size):
...
+            self._recorded_all_results.set()
 
+        # there is a choice here of whether to run
+        # the model thread while waiting for the next step call
+        # solved by moving the barrier to _record_logs to stop
+        # training after it ends rather than before it begins
+        self._ready_to_continue.wait()
+        self._step_counter = 0
+
+    def _generic_step(self, number_of_steps, config_key, action_fn, res_fn):
```
Comment (richardliaw) [inline python/ray/experimental/sgd/tf/tf_runner.py]: Can you document this method? It'll be very important to document the specific implementation for future reference.

**Comment by maximsmol** [comment]
Suspending work on this until an apparent tensorflow 2 memory leak gets fixed. Right now running even a trivial (mem_test.py) example with MultiWorker strategy causes OOM errors

**Comment by maximsmol** [comment]
Suspending work on this until an apparent tensorflow 2 memory leak gets fixed. Right now running even a trivial (mem_test.py) example with MultiWorker strategy causes OOM errors

https://github.com/tensorflow/tensorflow/issues/37057

**Comment by richardliaw** [comment]
Maybe let's close this PR for now and reopen it later.
